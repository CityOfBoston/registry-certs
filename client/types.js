// @flow

import type {
  FetchDeathCertificatesQuery,
  SearchDeathCertificatesQuery,
  LookupDeathCertificateOrderQuery,
} from './queries/graphql-types';

export type DeathCertificate = $NonMaybeType<
  $ArrayElement<
    $PropertyType<
      $PropertyType<FetchDeathCertificatesQuery, 'deathCertificates'>,
      'certificates'
    >
  >
>;

export type DeathCertificateSearchResults = $PropertyType<
  $PropertyType<SearchDeathCertificatesQuery, 'deathCertificates'>,
  'search'
>;

export type DeathCertificateOrder = $PropertyType<
  $PropertyType<LookupDeathCertificateOrderQuery, 'deathCertificates'>,
  'order'
>;
