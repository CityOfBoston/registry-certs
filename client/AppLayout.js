// @flow
/* eslint react/display-name: 0 */

import React, {
  type Element as ReactElement,
  type ChildrenArray as ReactChildrenArray,
} from 'react';

import headerHtml from '../templates/header.html';
import navigationHtml from '../templates/navigation.html';
import footerHtml from '../templates/footer.html';

import type { ClientDependencies } from './app';
import Nav, { type Props as NavProps, type LinkOptions } from './common/Nav';

type Props = {|
  children: ReactChildrenArray<ReactElement<*>>,
  navProps: NavProps,
|};

export default function AppLayout({ children, navProps }: Props) {
  // TODO(fin): remove wrapper <div> and return an array w/ React 16
  return (
    <div>
      <input
        type="checkbox"
        id="brg-tr"
        className="brg-tr"
        aria-hidden="true"
      />
      <nav
        className="nv-m"
        dangerouslySetInnerHTML={{ __html: navigationHtml }}
      />

      <div className="a11y--h" aria-live="polite" id="ariaLive" />

      <div className="mn mn--full mn--nv-s" style={{ zIndex: 2 }}>
        <input type="checkbox" id="s-tr" className="s-tr" aria-hidden="true" />
        <header
          className="h"
          role="banner"
          dangerouslySetInnerHTML={{ __html: headerHtml }}
        />

        <Nav {...navProps} />

        {children}
      </div>

      <footer
        className="ft"
        style={{ position: 'relative', zIndex: 2 }}
        dangerouslySetInnerHTML={{ __html: footerHtml }}
      />
    </div>
  );
}

export const wrapAppLayout = <P>(
  navLink: LinkOptions,
  render: (ClientDependencies, P) => ReactElement<*>
) => (dependencies: ClientDependencies, props: P) => (
  <AppLayout navProps={{ cart: dependencies.cart, link: navLink }}>
    {render(dependencies, props)}
  </AppLayout>
);
