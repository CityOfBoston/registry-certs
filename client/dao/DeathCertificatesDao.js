// @flow

import DataLoader from 'dataloader';
import type { LoopbackGraphql } from '../lib/loopback-graphql';
import type { DeathCertificate } from '../types';

import fetchDeathCertificates from '../queries/fetch-death-certificates';
import searchDeathCertificates from '../queries/search-death-certificates';
import lookupDeathCertificateOrder from '../queries/lookup-death-certificate-order';
import type {
  DeathCertificateSearchResults,
  DeathCertificateOrder,
} from '../types';

export type DeathCertificateCache = { [id: string]: DeathCertificate };

export default class DeathCertificatesDao {
  loopbackGraphql: LoopbackGraphql;
  loader: DataLoader<string, ?DeathCertificate>;

  cacheFullQuery: string;
  cacheByPage: { [page: number]: DeathCertificateSearchResults };

  constructor(loopbackGraphql: LoopbackGraphql) {
    this.loopbackGraphql = loopbackGraphql;

    // create new array shenanigans to get Flow to accept that we're not returning Errors
    this.loader = new DataLoader(ids =>
      fetchDeathCertificates(loopbackGraphql, ids).then(a => [...a])
    );
  }

  get(id: string): Promise<?DeathCertificate> {
    return this.loader.load(id);
  }

  async search(
    fullQuery: string,
    page: number
  ): Promise<DeathCertificateSearchResults> {
    if (fullQuery === this.cacheFullQuery) {
      const results = this.cacheByPage[page];
      if (results) {
        return results;
      }
    } else {
      this.cacheFullQuery = fullQuery;
      this.cacheByPage = {};
    }

    const { query, startYear, endYear } = this.parseQuery(fullQuery);

    const results = await searchDeathCertificates(
      this.loopbackGraphql,
      query,
      page,
      startYear,
      endYear
    );

    this.cacheByPage[page] = results;

    results.results.forEach(cert => {
      this.loader.prime(cert.id, cert);
    });

    return results;
  }

  async lookupOrder(
    id: string,
    contactEmail: string
  ): Promise<?DeathCertificateOrder> {
    return lookupDeathCertificateOrder(this.loopbackGraphql, id, contactEmail);
  }

  parseQuery(
    fullQuery: string
  ): { query: string, startYear: ?string, endYear: ?string } {
    // match a 4-digit year, and optionally a second 4-digit year with a hypen or en-dash
    const yearRegexp = /(\d{4})\s*[-â€“]?\s*(\d{4})?/;

    let query;
    let startYear;
    let endYear;

    const match = fullQuery.match(yearRegexp);
    if (match) {
      query = fullQuery.replace(yearRegexp, '').trim();
      startYear = match[1];
      endYear = match[2] || match[1];
    } else {
      query = fullQuery;
      startYear = null;
      endYear = null;
    }

    return {
      query,
      startYear,
      endYear,
    };
  }
}
