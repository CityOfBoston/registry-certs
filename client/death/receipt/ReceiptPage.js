// @flow

import React from 'react';
import Head from 'next/head';

import type { DeathCertificateOrder } from '../../types';

import {
  getDependencies,
  type ClientContext,
  type ClientDependencies,
} from '../../app';

type Props = {|
  order: ?DeathCertificateOrder,
|};

export default class ReceiptPage extends React.Component<Props> {
  static async getInitialProps(
    ctx: ClientContext,
    dependenciesForTest?: ClientDependencies
  ): Promise<Props> {
    const { query: { id, contactEmail }, res } = ctx;
    const { deathCertificatesDao } =
      dependenciesForTest || getDependencies(ctx);

    if (!id) {
      throw new Error('Missing id');
    }

    if (!contactEmail) {
      throw new Error('Missing contactEmail');
    }

    const order = await deathCertificatesDao.lookupOrder(id, contactEmail);

    if (!order && res) {
      res.statusCode = 404;
    }

    return {
      order,
    };
  }

  render() {
    const { order } = this.props;

    if (!order) {
      throw new Error('Order was not found');
    }

    const {
      id,
      date,
      shippingName,
      shippingCompanyName,
      shippingAddress1,
      shippingAddress2,
      shippingCity,
      shippingState,
      shippingZip,
      items,
      subtotal,
      serviceFee,
      total,
    } = order;

    return (
      <div className="b-ff">
        <Head>
          <title>Boston.gov — Death Certificates — Order #{order.id}</title>
        </Head>

        <div className="b-c b-c--hsm b-ff">
          <div className="lo m-b300 logo-holder">
            <span className="lo-l">
              <img
                src="https://patterns.boston.gov/images/public/logo.svg"
                alt="City of Boston"
                className="lo-i"
              />
            </span>
            <a
              className="print-button"
              href="javascript:void(0)"
              onClick={() => window.print()}
              style={{ fontStyle: 'italic' }}
            >
              Print this page
            </a>
          </div>

          <div className="b--g p-a300 br br-a200">
            <div className="sh sh--sm">
              <h1 className="sh-title">Thank you for your order</h1>
            </div>

            <div className="m-v300 t--cb">
              <div className="g">
                <div className="g--6">
                  <div className="m-t200">
                    <span className="receipt-title">Date:</span>{' '}
                    <span className="receipt-value">{date}</span>
                  </div>
                  <div className="m-t200">
                    <span className="receipt-title">Order #:</span>{' '}
                    <span className="receipt-value">{id}</span>
                  </div>
                </div>

                <div className="g--6">
                  <div className="m-t200 receipt-title">
                    Your shipping information:
                  </div>
                  <div className="receipt-value">
                    {shippingName}
                    <br />
                    {shippingCompanyName
                      ? [shippingCompanyName, <br key="br" />]
                      : null}
                    {shippingAddress1}
                    <br />
                    {shippingAddress2
                      ? [shippingAddress2, <br key="br" />]
                      : null}
                    {`${shippingCity}, ${shippingState} ${shippingZip}`}
                  </div>
                </div>
              </div>

              <div>
                <div className="m-t200 receipt-title">Your order:</div>
                <table>
                  <tbody>
                    {items.map(
                      ({ certificate, quantity, cost }) =>
                        certificate && (
                          <tr key={certificate.id}>
                            <td className="receipt-value">
                              {quantity} × Death certificate for{' '}
                              {certificate.firstName} {certificate.lastName} (#{certificate.id})
                            </td>
                            <td className="ta-r" style={{ fontWeight: 'bold' }}>
                              ${(cost / 100).toFixed(2)}
                            </td>
                          </tr>
                        )
                    )}

                    <tr>
                      <td className="receipt-value">Subtotal</td>
                      <td className="ta-r" style={{ fontWeight: 'bold' }}>
                        ${(subtotal / 100).toFixed(2)}
                      </td>
                    </tr>

                    <tr>
                      <td className="receipt-value">Card service fee*</td>
                      <td className="ta-r" style={{ fontWeight: 'bold' }}>
                        ${(serviceFee / 100).toFixed(2)}
                      </td>
                    </tr>

                    <tr className="total-row t--sans">
                      <td>Total</td>
                      <td className="ta-r">${(total / 100).toFixed(2)}</td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>

            <div className="t--cb m-v300">
              Your order will be shipped within 1–2 business days via USPS.
            </div>

            <div className="t--subinfo m-t300">
              * You are charged an extra service fee to pay for the cost of card
              processing. This fee goes directly to a third party, not the City
              of Boston.
            </div>
          </div>

          <div className="t--info m-v300 seal-holder">
            <img
              src="https://patterns.boston.gov/images/public/seal.svg"
              alt=""
              className="s-i"
              style={{ width: 75, height: 75 }}
            />

            <div className="m-l300">
              If you have any issues or questions, contact the Registry at{' '}
              <a href="tel:617-635-4175">617-635-4175</a> or email{' '}
              <a href="mailto:registry@boston.gov">registry@boston.gov</a>.
            </div>
          </div>
        </div>

        <style jsx>
          {`
            @media print {
              .print-button {
                display: none;
              }

              * {
                color: #000 !important;
              }

              .b-c--hsm {
                min-width: 100% !important;
                padding: 0 !important;
              }

              .b--g {
                background-color: white !important;
              }
            }

            .logo-holder {
              display: flex;
              justify-content: space-between;
            }

            .receipt-title {
              font-weight: bold;
              line-height: 1.4;
            }

            .receipt-value {
              font-style: italic;
              line-height: 1.4;
            }

            table {
              width: 100%;
              line-height: 1.4;
            }

            .total-row {
              font-size: 2rem;
              text-transform: uppercase;
              font-weight: bold;
            }

            .seal-holder {
              display: flex;
              align-items: center;
            }
          `}
        </style>
      </div>
    );
  }
}
