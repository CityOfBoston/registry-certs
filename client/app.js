// @flow
/* global Stripe */

import { useStrict } from 'mobx';
import Router from 'next/router';

import type { Context as NextContext } from 'next';

import RouterListener from './lib/RouterListener';

import makeLoopbackGraphql, {
  type LoopbackGraphql,
} from './lib/loopback-graphql';

import type { RequestAdditions } from '../server/lib/request-additions';

import Cart from './store/Cart';
import OrderProvider from './store/OrderProvider';

import DeathCertificatesDao from './dao/DeathCertificatesDao';
import CheckoutDao from './dao/CheckoutDao';

export type ClientContext = NextContext<RequestAdditions>;

export type ClientDependencies = {
  stripe: ?StripeInstance,
  loopbackGraphql: LoopbackGraphql,
  cart: Cart,
  orderProvider: OrderProvider,
  deathCertificatesDao: DeathCertificatesDao,
  checkoutDao: CheckoutDao,
};

let browserInited = false;
let browserDependencies: ClientDependencies;

// Browser-only setup
export function initBrowser() {
  if (browserInited) {
    return;
  }

  browserInited = true;

  useStrict(true);

  const routerListener = new RouterListener();
  routerListener.attach(Router);
}

// Works on both server and browser. Memoizes on browser, so these dependencies
// will share state across pages.
//
// Typically called from a wrapper function or controller component to inject
// these into a component's props.
export function getDependencies(ctx?: ClientContext): ClientDependencies {
  if (process.browser && browserDependencies) {
    return browserDependencies;
  }

  // req will exist only when this function is called for getInitialProps on the
  // server.
  const req = ctx && ctx.req;

  const stripe =
    typeof Stripe !== 'undefined'
      ? Stripe((window.__NEXT_DATA__ || {}).stripePublishableKey || '')
      : null;
  const loopbackGraphql = makeLoopbackGraphql(req);
  const deathCertificatesDao = new DeathCertificatesDao(loopbackGraphql);
  const checkoutDao = new CheckoutDao(loopbackGraphql, stripe);
  const cart = new Cart();
  const orderProvider = new OrderProvider();

  if (process.browser) {
    cart.attach(window.localStorage, deathCertificatesDao);
    orderProvider.attach(window.localStorage);
  }

  const dependencies: ClientDependencies = {
    stripe,
    cart,
    orderProvider,
    deathCertificatesDao,
    checkoutDao,
    loopbackGraphql,
  };

  if (process.browser) {
    browserDependencies = dependencies;
  }

  return dependencies;
}
