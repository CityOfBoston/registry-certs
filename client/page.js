// @flow

import React from 'react';
import type { Context } from 'next';
import { useStrict } from 'mobx';
import Router from 'next/router';

import Cart from './store/Cart';
import makeLoopbackGraphql from './loopback-graphql';
import type { LoopbackGraphql } from './loopback-graphql';
import type { RequestAdditions } from '../server/request-additions';

import DeathCertificatesDao from './dao/DeathCertificatesDao';
import RouterListener from './RouterListener';

// Higher-order component for a Page in our app.

// Pass this mixin a function that requires the component class and returns it.
// require is necessary rather than import since imports are hoisted to the
// beginning of files, and we need to rehydrate before any Glamor "css"
// statements are processed.

export type ClientDependencies = {
  cart: Cart,
  deathCertificatesDao: DeathCertificatesDao,
  loopbackGraphql: LoopbackGraphql,
};

let browserInited = false;
let browserDependencies: ClientDependencies;

function maybeInitBrowserLibraries() {
  if (browserInited || !process.browser) {
    return;
  }

  browserInited = true;

  useStrict(true);

  const routerListener = new RouterListener();
  routerListener.attach(Router);
}

function makeDependencies(req: ?RequestAdditions): ClientDependencies {
  if (process.browser && browserDependencies) {
    return browserDependencies;
  }

  const loopbackGraphql = makeLoopbackGraphql(req);
  const deathCertificatesDao = new DeathCertificatesDao(loopbackGraphql);

  const cart = new Cart();
  if (process.browser) {
    cart.attach(window.localStorage, deathCertificatesDao);
  }

  const dependencies: ClientDependencies = {
    cart,
    deathCertificatesDao,
    loopbackGraphql,
  };

  if (process.browser) {
    browserDependencies = dependencies;
  }

  return dependencies;
}

export default <OP, P: $Subtype<Object>, S>(
  Component: Class<React.Component<OP, P, S>>
): Class<React.Component<void, P, void>> => {
  maybeInitBrowserLibraries();

  return class Page extends React.Component {
    props: P;
    loopbackGraphql: LoopbackGraphql = makeLoopbackGraphql();

    static getInitialProps(ctx: Context<*>) {
      const { req } = ctx;

      const dependencies = makeDependencies(req);

      if (typeof Component.getInitialProps === 'function') {
        return Component.getInitialProps(ctx, dependencies);
      } else {
        return {};
      }
    }

    render() {
      const dependencies = makeDependencies();
      return <Component {...dependencies} {...this.props} />;
    }
  };
};
