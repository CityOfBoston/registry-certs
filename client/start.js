// @flow

// Utility file for automatically calling initialization functions when page
// modules are run. We delegate to a separate module so that that module can be
// tested without automatically running on import.
//
// Note that this is called for every page on the browser, but only once on the
// server when the app is first interpreted on startup.
//
// This is typically imported and then the page component is require'd in. Using
// require enforces an execution order such that initBrowser can e.g. hydrate
// CSS-in-JS libraries before the "css" method is called during component module
// execution.

import { initBrowser } from './app.js';

if (process.browser) {
  initBrowser();
}
