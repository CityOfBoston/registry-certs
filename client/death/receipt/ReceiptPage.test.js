// @flow
import DeathCertificatesDao from '../../dao/DeathCertificatesDao';

import ReceiptPage from './ReceiptPage';

import { TYPICAL_ORDER } from '../../../fixtures/client/death-certificate-orders';

jest.mock('../../dao/DeathCertificatesDao');

describe('getInitialProps', () => {
  let deathCertificatesDao;

  beforeEach(() => {
    deathCertificatesDao = new DeathCertificatesDao((null: any));
  });

  it('loads the order passed in query', async () => {
    deathCertificatesDao.lookupOrder.mockReturnValue(TYPICAL_ORDER);

    const initialProps = await ReceiptPage.getInitialProps(
      ({
        query: { id: 'RG-DC201801-360926', contactEmail: 'nancy@mew.org' },
      }: any),
      ({ deathCertificatesDao }: any)
    );

    expect(deathCertificatesDao.lookupOrder).toHaveBeenCalledWith(
      'RG-DC201801-360926',
      'nancy@mew.org'
    );
    expect(initialProps).toMatchSnapshot();
  });

  it('handles a 404', async () => {
    deathCertificatesDao.get.mockReturnValue(null);

    const initialProps = await ReceiptPage.getInitialProps(
      ({
        query: { id: 'RG-DC201801-360926', contactEmail: 'nancy@mew.org' },
      }: any),
      ({ deathCertificatesDao }: any)
    );

    expect(deathCertificatesDao.lookupOrder).toHaveBeenCalledWith(
      'RG-DC201801-360926',
      'nancy@mew.org'
    );
    expect(initialProps).toMatchSnapshot();
  });
});
