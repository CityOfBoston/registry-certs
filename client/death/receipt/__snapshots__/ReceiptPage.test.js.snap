// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getInitialProps handles a 404 1`] = `
Object {
  "order": undefined,
}
`;

exports[`getInitialProps loads the order passed in query 1`] = `
Object {
  "order": Object {
    "certificateCost": 1400,
    "contactEmail": "nancy@mew.org",
    "contactName": "Nancy Whitehead",
    "contactPhone": "5551234567",
    "date": "1/8/2018 2:05PM",
    "id": "RG-DC201801-360926",
    "items": Array [
      Object {
        "certificate": Object {
          "firstName": "JAMES",
          "id": "640768",
          "lastName": "HOWLETT (LOGAN)",
        },
        "cost": 9800,
        "quantity": 7,
      },
      Object {
        "certificate": Object {
          "firstName": "BRUCE",
          "id": "639896",
          "lastName": "BANNER",
        },
        "cost": 1400,
        "quantity": 1,
      },
    ],
    "serviceFee": 266,
    "shippingAddress1": "Avengers Tower",
    "shippingAddress2": "",
    "shippingCity": "New York",
    "shippingCompanyName": "US Avengers",
    "shippingName": "Squirrel Girl",
    "shippingState": "NY",
    "shippingZip": "10001",
    "subtotal": 11200,
    "total": 11466,
  },
}
`;
